client/src/explorer_common/explorer_components.js

```diff
@@ -167,24 +167,34 @@ export const ExplorerNode = ({
           onClick={noExpand ? null : () => onClickExpand(node)}
         >
           <div className="ExplorerRow">
-            {_.map(column_defs, ({ id, width, get_val, val_display }, ix) => (
-              <div
-                key={id}
-                className="ExplorerRow__Cell"
-                style={
-                  ix === 0
-                    ? {
-                        ...computed_col_styles[id],
-                        flex: `1 0 ${width - depth * INDENT_SIZE}px`,
-                      }
-                    : computed_col_styles[id]
-                }
-              >
-                {_.isFunction(val_display)
-                  ? val_display(get_val(node))
-                  : get_val(node)}
-              </div>
-            ))}
+            {_.map(
+              column_defs,
+              (
+                { id, width, get_val, val_display, first_node_m2m_display },
+                ix
+              ) => (
+                <div
+                  key={id}
+                  className="ExplorerRow__Cell"
+                  style={
+                    ix === 0
+                      ? {
+                          ...computed_col_styles[id],
+                          flex: `1 0 ${width - depth * INDENT_SIZE}px`,
+                        }
+                      : computed_col_styles[id]
+                  }
+                >
+                  {node.data.is_m2m &&
+                  _.isFunction(first_node_m2m_display) &&
+                  depth === 0
+                    ? first_node_m2m_display(get_val(node))
+                    : _.isFunction(val_display)
+                    ? val_display(get_val(node))
+                    : get_val(node)}
```

Comment:
IMO, `explorer_components.js` doesn't need to know about these weird special case displays for the first node in a many-to-many hierarchy. Bit of a concept leak. A developer just reading this code might have no idea what a `first_node_m2m_display`. In fact, the explorers using it could all go away or be refactored, and then it'd be a totally mysterious orphaned name inside this module.

I'd suggest `explorer_components.js` could pass `node` as a second argument to `val_display`. Then consumers of the explorer components could, optionally, implement additional `val_display` logic based on things like node depth, without the common explorer components needing to know anything about it.